router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
```

Comment:
Damn, I think this might have a small race. If `roundTrip()` gets a `dialErr` trying to hit one backend, then it might already be moving on to another request when the old request gets canceled.

I originally had some more coordination that trickled down into `roundTrip()` but thought it might be unnecessary. Now I'm not sure if there's a way to avoid that.